packages/oui-file/src/file.controller.js

```diff
@@ -0,0 +1,204 @@
+import find from "lodash/find";
+import isEmpty from "lodash/isEmpty";
+import remove from "lodash/remove";
+
+export default class {
+    constructor ($attrs, $compile, $element, $filter, $scope, $timeout, $window, ouiFileConfiguration) {
+        "ngInject";
+
+        this.$attrs = $attrs;
+        this.$compile = $compile;
+        this.$element = $element;
+        this.$filter = $filter;
+        this.$scope = $scope;
+        this.$timeout = $timeout;
+        this.$window = $window;
+
+        this.translations = ouiFileConfiguration.translations;
+        this.units = this.$filter("orderBy")(ouiFileConfiguration.units, "-size");
+    }
+
+    checkFileValidity (file) {
+        if (file) {
+            file.errors = {};
+
+            // Check maxsize
+            if (this.maxsize && file.size > this.maxsize) {
+                file.errors.maxsize = true;
+            }
+
+            // Set form validation
+            if (this.form && this.form[this.name]) {
+                this.form[this.name].$setValidity("maxsize", !file.errors.maxsize);
+                this.form[this.name].$setDirty();
+            }
+
+            // Clean errors
+            if (isEmpty(file.errors)) {
+                delete file.errors;
+            }
+        }
+
+        return file;
+    }
+
+    loadFilePreview (file) {
+        // Load preview only if image
+        if (this.preview && !file.errors && file.type && file.type.search(/^image\//) !== -1) {
+            file.loading = true;
+
+            file.reader = new this.$window.FileReader();
+            file.reader.readAsDataURL(file);
+            file.reader.onload = () => {
+                file.preview = `url("${file.reader.result}")`;
+
+                // Clean file object
+                delete file.loading;
+                delete file.reader;
+
+                this.$scope.$apply();
+            };
+        }
+
+        return file;
+    }
+
+    addFile (file) {
+        this.model = [this.checkFileValidity(file)];
+        this.onSelect({ modelValue: this.model });
+        this.$scope.$apply();
+
+        // Set back focus on fake selector
+        this.$element[0].querySelector(".oui-file-selector__label").focus();
+    }
+
+    addFiles (files) {
+        if (!this.model) {
+            this.model = [];
+        }
+
+        if (angular.isArray(files)) {
+            files.forEach((file) => {
+                // Check for duplicate before adding
+                if (!find(this.model, (item) => file.name === item.name)) {
+                    this.checkFileValidity(file);
+                    this.loadFilePreview(file);
+                    this.model.push(file);
+                }
+            });
+
+            this.onSelect({ modelValue: this.model });
+            this.$scope.$apply();
+        }
+    }
+
+    removeFile (file) {
+        if (angular.isArray(this.model)) {
+            remove(this.model, (item) => item === file);
```

Comment:
So a `file` is a string ? Wouldn't it make more sense to have it being an object with a `path` property ?

```js
{
  path: '...'
  isReadingPreview: true,
  preview: [...]
}
```

Reply:
No, a file is a file object. What make you think it's a string ?